patroni/scripts/barman_recover.py

```diff
@@ -0,0 +1,494 @@
+#!/usr/bin/python3
+
+"""Restore a Barman backup to the local node through ``pg-backup-api``.
+
+This script can be used both as a custom bootstrap method, and as a custom
+create replica method. Check the output of ``--help`` to understand the
+parameters supported by the script. ``--datadir`` is a special parameter and it
+is automatically filled by Patroni in both cases.
+
+It requires that you have previously configured a Barman server, and that you
+have ``pg-backup-api`` configured and running in the same host as Barman.
+
+Refer to :class:`ExitCode` for possible exit codes of this script.
+"""
+from argparse import ArgumentParser
+from enum import IntEnum
+import json
+import logging
+import ssl
+import sys
+import time
+from typing import Any, Callable, Optional, Tuple, Type, Union, TYPE_CHECKING
+from urllib.error import HTTPError, URLError
+from urllib.parse import urljoin
+from urllib.request import Request, urlopen
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    from http.client import HTTPResponse
+
+
+class ExitCode(IntEnum):
+    """Possible exit codes of this script.
+
+    :cvar RECOVERY_DONE: backup was successfully restored.
+    :cvar RECOVERY_FAILED: recovery of the backup faced an issue.
+    :cvar API_NOT_OK: ``pg-backup-api`` status is not ``OK``.
+    :cvar BARMAN_SERVER_DOES_NOT_EXIST: there is no such Barman server
+        configured in the Barman host.
+    :cvar HTTP_REQUEST_ERROR: an error has occurred during a request to the
+        ``pg-backup-api``.
+    :cvar HTTP_RESPONSE_MALFORMED: ``pg-backup-api`` returned a bogus response.
+    """
+
+    RECOVERY_DONE = 0
+    RECOVERY_FAILED = 1
+    API_NOT_OK = 2
+    BARMAN_SERVER_DOES_NOT_EXIST = 3
+    HTTP_REQUEST_ERROR = 4
+    HTTP_RESPONSE_MALFORMED = 5
+
+
+class RetriesExceeded(Exception):
+    """Maximum number of retries exceeded."""
+
+
+def retry(times_attr: str, retry_wait_attr: str,
+          exceptions: Union[Type[Exception], Tuple[Type[Exception], ...]]) \
```

Comment:
Do we need to have this flexibility? It is used only in this file.

Reply:
Regarding attribute names? I may remove that and add a note, not a problem :)